### PR TITLE
QData[RValue]

### PR DESCRIPTION
**Pull request by alissapajer**
Depends on #3777 and #3780.

### CODE REVIEW

File: blueeyes/src/main/scala/quasar/precog/common/QDataRValue.scala [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+  def getArrayAt(ac: ArrayCursor): RValue = ac.values(ac.index.toInt)
+  def stepArray(ac: ArrayCursor): ArrayCursor = ac.copy(index = ac.index + Natural.one)
+
+  type NascentArray = List[RValue]
+
+  def prepArray: NascentArray = List[RValue]()
+  def pushArray(a: RValue, na: NascentArray): NascentArray = a +: na // prepend
+  def makeArray(na: NascentArray): RValue = if (na.isEmpty) CEmptyArray else RArray(na.reverse)
+
+  final case class ObjectCursor(index: Natural, values: List[(String, RValue)])
```
Comment (wemrysi) [inline blueeyes/src/main/scala/quasar/precog/common/QDataRValue.scala]: Same argument here as with `ArrayCursor`.

File: blueeyes/src/main/scala/quasar/precog/common/QDataRValue.scala [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    case _ => error(s"Expected `RArray` or `CEmptyArray`. Received $a")
+  }
+  def hasNextArray(ac: ArrayCursor): Boolean = ac.values.length > ac.index.toInt
+  def getArrayAt(ac: ArrayCursor): RValue = ac.values(ac.index.toInt)
+  def stepArray(ac: ArrayCursor): ArrayCursor = ac.copy(index = ac.index + Natural.one)
+
+  type NascentArray = List[RValue]
+
+  def prepArray: NascentArray = List[RValue]()
+  def pushArray(a: RValue, na: NascentArray): NascentArray = a +: na // prepend
```
Comment (wemrysi) [inline blueeyes/src/main/scala/quasar/precog/common/QDataRValue.scala]: Minor, `::` is more idiomatic when prepending to a list.

File: blueeyes/src/main/scala/quasar/precog/common/QDataRValue.scala [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+    case RObject(obj) => ObjectCursor(Natural.zero, obj.toList)
+    case CEmptyObject => ObjectCursor(Natural.zero, List[(String, RValue)]())
+    case _ => error(s"Expected `Data.Obj`. Received $a")
+  }
+  def hasNextObject(ac: ObjectCursor): Boolean = ac.values.length > ac.index.toInt
+  def getObjectKeyAt(ac: ObjectCursor): String = ac.values(ac.index.toInt)._1
+  def getObjectValueAt(ac: ObjectCursor): RValue = ac.values(ac.index.toInt)._2
+  def stepObject(ac: ObjectCursor): ObjectCursor = ac.copy(index = ac.index + Natural.one)
+
+  type NascentObject = List[(String, RValue)]
```
Comment (wemrysi) [inline blueeyes/src/main/scala/quasar/precog/common/QDataRValue.scala]: I think we could just use a `Map[String, RValue]` here and avoid the list traversals altogether. We lose all insert order when we convert to `Map` at the end, so `List` doesn't appear to gain us anything.

File: blueeyes/src/main/scala/quasar/precog/common/QDataRValue.scala [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+  }
+  def makeOffsetTime(l: OffsetTime): RValue = COffsetTime(l)
+
+  def getInterval(a: RValue): DateTimeInterval = a match {
+    case CInterval(value) => value
+    case _ => error(s"Expected `CInterval`. Received $a")
+  }
+  def makeInterval(l: DateTimeInterval): RValue = CInterval(l)
+
+  final case class ArrayCursor(index: Natural, values: List[RValue])
```
[the remark this answers is not in the document]
Comment (wemrysi) [reply blueeyes/src/main/scala/quasar/precog/common/QDataRValue.scala]: If the contract is that `stepArray` should only be called when `hasNextArray` returns `true` then we could actually use `_.tail` to implement `stepArray`.

File: blueeyes/src/main/scala/quasar/precog/common/QDataRValue.scala (hunk 3 above)
Comment (alissapajer) [reply blueeyes/src/main/scala/quasar/precog/common/QDataRValue.scala]: That's a good point. There was a reason I used `List` but I can't remember what it was, so I'll make the `Map` change and see what falls out :)

**Comment by alissapajer** [comment]
@wemrysi thanks for the suggestions! The `QData[RValue]` implementation largely mirrored the `QData[Data]` implementation, so whatever changes I make, I'll make them over in `QData[Data]` as well, where appropriate.

File: blueeyes/src/main/scala/quasar/precog/common/QDataRValue.scala (hunk 3 above)
Comment (alissapajer) [reply blueeyes/src/main/scala/quasar/precog/common/QDataRValue.scala]: I figured out why I had `List` here. In `Data.Obj` we store the values in a `ListMap[String, Data]` and so order must be preserved in that case. But in `RValue` we use `Map[String, RValue]` so we can indeed use `Map` here!

File: blueeyes/src/main/scala/quasar/precog/common/QDataRValue.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (alissapajer) [reply blueeyes/src/main/scala/quasar/precog/common/QDataRValue.scala]: I wonder if we have enough or long enough arrays/objects in our integration tests for this fix to make them even faster!

File: blueeyes/src/main/scala/quasar/precog/common/QDataRValue.scala (hunk 3 above)
Comment (wemrysi) [reply blueeyes/src/main/scala/quasar/precog/common/QDataRValue.scala]: Gotcha. We could use `ListMap` in the `Data` instance directly as it preserves insertion order.

**Comment by alissapajer** [comment]
@wemrysi ready for another look.

**Comment by wemrysi** [comment]
@alissapajer Would you mind adding an implicit `QData` instance to the companion objects of `Data` and `RValue` so it may be found automatically by implicit search?

**Comment by alissapajer** [comment]
@wemrysi implicits added!

**Pull request merged**